### ISSUE TITLE
Can't get the result of the professional----> bad AP

### ISSUE THREAD

**Issue by Lydiagugugaga** [yeonsikch/Unbiased-Teacher#2]
Hello! Thanks for your work.

Since I only have access to 1 GPU, I am using smaller batch size (label: 6images, unlabel: 6 images).

Finally, I get worse Box AP results. AP=8 (for coco standard, 1%sup, base_LR=0.01)
The Loss_box_reg and the Loss_box_reg_pseudo could not be declined, and the total loss is about 0.4 to 0.5 all the time.

How can I reproduce the result of Unbaised-Teacher?

I'm looking forward to your reply!

**Comment by yeonsikch** [comment]
Hi.

Did you learn until end of end?
And Batch size is important because of batch normalization.

I think that you have to ask this paper's author.
I'm not author and I don't exactly know this issue. sorry.

this is author's github.
https://github.com/facebookresearch/unbiased-teacher

Thank you.

**Comment by Lydiagugugaga** [comment]
Thanks very much for your reply. 

I have asked the author about the previous problem before,  but  there was still no clear answer, maybe modifying the unsup_loss_weight and so on, it didin't work in the end. It is estimated that the batchsize has a great influence, so now I choose the smaller dataset to reduce the occupation of GPU memory.

**Can I ask you a new question?**
Because of the limitation of GPU memory, I tried to use the Pascal VOC (smaller than COCO) as my dataset. My device is like yours, I use one GPU for training (3090, Batchsize: label 8, unlabel 8). But the training time is to long, I need to spend 2 days to complete this project. Here is my config( **the same as the author's config**, I am not sure which parameter I can modify to improve the speed):

> ROI_HEADS:
>     NAME: "StandardROIHeadsPseudoLab"
>     LOSS: "FocalLoss"
>     NUM_CLASSES: 20
> INPUT:
>   MIN_SIZE_TRAIN: (480, 512, 544, 576, 608, 640, 672, 704, 736, 768, 800)
>   MIN_SIZE_TEST: 600
> SOLVER:
>   LR_SCHEDULER_NAME: "WarmupMultiStepLR"
>   STEPS: (179990, 179995)
>   MAX_ITER: 180000
>   IMG_PER_BATCH_LABEL: 16
>   IMG_PER_BATCH_UNLABEL: 16
>   BASE_LR: 0.01
> DATALOADER:
>   SUP_PERCENT: 100.0
> DATASETS:
>   CROSS_DATASET: True
>   TRAIN: ("voc_2012_trainval",)
>   TRAIN_LABEL: ("voc_2007_trainval",)
>   TRAIN_UNLABEL: ("voc_2012_trainval",)
>   TEST: ("voc_2007_test",)
> SEMISUPNET:
>   Trainer: "ubteacher"
>   BBOX_THRESHOLD: 0.7
>   TEACHER_UPDATE_ITER: 1
>   BURN_UP_STEP: 30000
>   EMA_KEEP_RATE: 0.999
>   UNSUP_LOSS_WEIGHT: 2.0
> TEST:
>   EVAL_PERIOD: 3000
>   EVALUATOR: "COCOeval"

I want to ask you about improving the training speed for advice, because I see that you also use one 3090 for experimentation.

I don’t have much experience in deep learning model training, I hope I can get your guidance.

**Comment by yeonsikch** [comment]
Hi.

This is my conf file for  10% labeled SS-OD.
But your config file have a `SUP_PERCENT: 100.0` 
This mean is you learn Supervised Learning via all labeled data.
This is not Semi-Supervised Learning.
And your `IMG_PER_BATCH_LABEL:16` and `IMG_PER_BATCH_UNLABEL:16` mean your batchsize is `label:unlabel = 16:16`
If this run, fix this size.

And I can recommend you about improving the training speed.
but this may so difficult.

you can NVIDIA/APEX-amp (https://github.com/NVIDIA/apex)
or pytorch-amp (https://pytorch.org/docs/stable/amp.html)

if you wanna this code, you have to modify some code.
So, this is so difficult.

Good Luck.

```
_BASE_: "../Base-RCNN-FPN.yaml"
MODEL:
  META_ARCHITECTURE: "TwoStagePseudoLabGeneralizedRCNN"
  WEIGHTS: "detectron2://ImageNetPretrained/MSRA/R-50.pkl"
  MASK_ON: False
  RESNETS:
    DEPTH: 50
  PROPOSAL_GENERATOR:
    NAME: "PseudoLabRPN"
  RPN:
    POSITIVE_FRACTION: 0.25
    LOSS: "CrossEntropy"
  ROI_HEADS:
    NAME: "StandardROIHeadsPseudoLab"
    LOSS: "FocalLoss"
    NUM_CLASSES: 15
SOLVER:
  LR_SCHEDULER_NAME: "WarmupMultiStepLR"
  STEPS: (179990, 179995)
  MAX_ITER: 180000
  IMG_PER_BATCH_LABEL: 4
  IMG_PER_BATCH_UNLABEL: 4
  BASE_LR: 0.01
DATALOADER:
  SUP_PERCENT: 10.0
  RANDOM_DATA_SEED: 1
DATASETS:
  CROSS_DATASET: False
  TRAIN: ("vtt_train",)
  TEST: ("vtt_val",)
  TRAIN_LABEL: ("vtt_train",)
  TRAIN_UNLABEL: ("vtt_train",)
SEMISUPNET:
  Trainer: "ubteacher"
  BBOX_THRESHOLD: 0.7
  TEACHER_UPDATE_ITER: 1
  BURN_UP_STEP: 2000
  EMA_KEEP_RATE: 0.9996
  UNSUP_LOSS_WEIGHT: 2.0
TEST:
  EVAL_PERIOD: 1000
  EVALUATOR: "COCOeval"
```

**Comment by Lydiagugugaga** [comment]
Thanks very much! 
About  `RANDOM_DATA_SEED: 1`, do I need to generate a new file which is similar to the /dataseed/COCO_supervision.txt?

**Comment by yeonsikch** [comment]
As i know, you can ignore it.
Just keep it.

**Comment by Lydiagugugaga** [comment]
Thank you very much! 
I'm so happy that I can have this chance to ask you for some advice here!

**Comment by yeonsikch** [comment]
I'm so really happy too that I can talk to you on this issue page.

btw, can I send you an e-mail?

**Comment by Lydiagugugaga** [comment]
Of course! My emai: 779648500@qq.com
                                20162299@cqu.edu.cn
Both OK!

**Comment by Lydiagugugaga** [comment]
Hello! 
I have already modified my config. `SUP_PERCENT: 10.0` 
I tried to train pascal voc (3090, Batchsize: label 8, unlabel 8), but I could only train until the 34999th , after which I burst the GPU memory. 

> copypaste: AP,AP50,AP75
  copypaste: 51.1172,79.4793,55.6608

My  `BURN_UP_STEP: 30000`. 
Is this result just the result of supervised training？
I want to try to continue training with a batchsize of 6 by using the` model_0034999.pth `, because of the limitation of GPU memory.

**Comment by yeonsikch** [comment]
Hi.
I've been busy these days, so I haven't been able to send an e-mail.

First, I think that you have to batch size 4 /4. Because your 8 / 8 is available in Burn-in stage. Burn-in stage use only Supervised label = 8. But, in Teacher - Student stage, you use label 8 + unlabeled 8 = 16. So, If your gpu memory is full in Burn-in stage. You have to batch size to set half size.

But, I don't know why you could only train until the 34999th instead of `BURN_UP_STEP: 30000`.
I recommend you check the config file.

And show me your last AP table about both Teacher and Student.
(p. s. First AP table is Teacher table, second AP table is Student table.)

**Comment by Lydiagugugaga** [comment]
Thanks very much for your reply!
That is my last AP which I sent before. I just evaluate this saved model.

My fist AP is here:
> AP,AP50,AP75
> 47.4385,76.5950,51.3178

I found my remaining training time suddenly add one day after the 30000th iteration. And until 36755 iterations it is out of GPU memory.

I'll try to batch size 4 /4 for resume this saved model for continue training. 

This is my config:

> _BASE_: "../Base-RCNN-FPN.yaml"
> MODEL:
>   META_ARCHITECTURE: "TwoStagePseudoLabGeneralizedRCNN"
>   WEIGHTS: "detectron2://ImageNetPretrained/MSRA/R-50.pkl"
>   MASK_ON: False
>   RESNETS:
>     DEPTH: 50
>   PROPOSAL_GENERATOR:
>     NAME: "PseudoLabRPN"
>   RPN:
>     POSITIVE_FRACTION: 0.25
>     LOSS: "CrossEntropy"
>   ROI_HEADS:
>     NAME: "StandardROIHeadsPseudoLab"
>     LOSS: "FocalLoss"
>     NUM_CLASSES: 20
> INPUT:
>   MIN_SIZE_TRAIN: (480, 512, 544, 576, 608, 640, 672, 704, 736, 768, 800)
>   MIN_SIZE_TEST: 600
> SOLVER:
>   LR_SCHEDULER_NAME: "WarmupMultiStepLR"
>   STEPS: (179990, 179995)
>   MAX_ITER: 180000
>   IMG_PER_BATCH_LABEL: 8
>   IMG_PER_BATCH_UNLABEL: 8
>   BASE_LR: 0.01
> DATALOADER:
>   SUP_PERCENT: 10.0
>   RANDOM_DATA_SEED: 1
> DATASETS:
>   CROSS_DATASET: True
>   TRAIN: ("voc_2012_trainval",)
>   TRAIN_LABEL: ("voc_2007_trainval",)
>   TRAIN_UNLABEL: ("voc_2012_trainval",)
>   TEST: ("voc_2007_test",)
> SEMISUPNET:
>   Trainer: "ubteacher"
>   BBOX_THRESHOLD: 0.7
>   TEACHER_UPDATE_ITER: 1
>   BURN_UP_STEP: 30000
>   EMA_KEEP_RATE: 0.996
>   UNSUP_LOSS_WEIGHT: 2.0
> TEST:
>   EVAL_PERIOD: 3000
>   EVALUATOR: "COCOeval"

**Comment by yeonsikch** [comment]
OK. Good luck to you!
